File: src-tauri/src/contracts/parser.rs
// SPDX-License-Identifier: BUSL-1.1

//! Server-side relecture of contract YAML/JSON.
//!
//! Defense-in-depth: even if the frontend validates, contracts on disk can be
//! edited externally. Same invariants as `src/lib/contracts/parser.ts`.

use std::collections::HashSet;

use regex::Regex;
use serde::Deserialize;
use thiserror::Error;

use super::{AllowedValue, Contract, Rule};

#[derive(Debug, Error)]
pub enum ContractError {
    #[error("invalid contract: {0}")]
    Invalid(String),
    #[error("invalid YAML: {0}")]
    Yaml(#[from] serde_yml::Error),
    #[error("invalid JSON: {0}")]
    Json(#[from] serde_json::Error),
}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Format {
    Yaml,
    Json,
    Auto,
}

pub fn parse_contract(source: &str, format: Format) -> Result<Contract, ContractError> {
    let trimmed = source.trim();
    if trimmed.is_empty() {
        return Err(ContractError::Invalid("source is empty".into()));
    }

    let contract: Contract = match format {
        Format::Json => serde_json::from_str(source)?,
        Format::Yaml => serde_yml::from_str(source)?,
        Format::Auto => {
            if trimmed.starts_with('{') || trimmed.starts_with('[') {
                match serde_json::from_str(source) {
                    Ok(c) => c,
                    Err(_) => serde_yml::from_str(source)?,
                }
            } else {
                serde_yml::from_str(source)?
            }
        }
    };

    validate(&contract)?;
    Ok(contract)
}

fn validate(contract: &Contract) -> Result<(), ContractError> {
    let ident = ident_regex();
    let duration = duration_regex();

    if !ident.is_match(&contract.name) {
        return Err(ContractError::Invalid(format!(
            "$.name: must match [A-Za-z_][A-Za-z0-9_]* (got {:?})",
            contract.name
        )));
    }
    if contract.version != 1 {
        return Err(ContractError::Invalid(format!(
            "$.version: unsupported version {} (expected 1)",
            contract.version
        )));
    }
    if contract.target.connection.is_empty() {
        return Err(ContractError::Invalid("$.target.connection: required".into()));
    }
    if contract.target.table.is_empty() {
        return Err(ContractError::Invalid("$.target.table: required".into()));
    }
    if contract.rules.is_empty() {
        return Err(ContractError::Invalid("$.rules: must be non-empty".into()));
    }

    let mut ids: HashSet<&str> = HashSet::new();
    for (i, rule) in contract.rules.iter().enumerate() {
        let path = format!("$.rules[{}]", i);
        let id = rule.id();
        if !ident.is_match(id) {
            return Err(ContractError::Invalid(format!(
                "{}.id: must match [A-Za-z_][A-Za-z0-9_]* (got {:?})",
                path, id
            )));
        }
        if !ids.insert(id) {
            return Err(ContractError::Invalid(format!(
                "{}.id: duplicate id {:?}",
                path, id
            )));
        }
        validate_rule(rule, &path, &duration)?;
    }

    Ok(())
}

fn validate_rule(rule: &Rule, path: &str, duration: &Regex) -> Result<(), ContractError> {
    match rule {
        Rule::NotNullPct {
            threshold_min_pct, ..
        } => {
            if !threshold_min_pct.is_finite()
                || *threshold_min_pct < 0.0
                || *threshold_min_pct > 100.0
            {
                return Err(ContractError::Invalid(format!(
                    "{}.threshold_min_pct: must be in [0, 100]",
                    path
                )));
            }
        }
        Rule::RegexMatch { pattern, .. } => {
            Regex::new(pattern).map_err(|e| {
                ContractError::Invalid(format!("{}.pattern: invalid regex: {}", path, e))
            })?;
        }
        Rule::LengthRange { min, max, .. } => check_int_range(*min, *max, path)?,
        Rule::NumericRange { min, max, .. } => check_float_range(*min, *max, path)?,
        Rule::DateRange {
            min,
            max,
            max_age,
            ..
        } => {
            if min.is_none() && max.is_none() && max_age.is_none() {
                return Err(ContractError::Invalid(format!(
                    "{}: provide at least one of min, max, max_age",
                    path
                )));
            }
            if let Some(age) = max_age {
                if !duration.is_match(age) {
                    return Err(ContractError::Invalid(format!(
                        "{}.max_age: must be <number><ms|s|m|h|d>",
                        path
                    )));
                }
            }
        }
        Rule::AllowedValues { values, .. } => {
            if values.is_empty() {
                return Err(ContractError::Invalid(format!(
                    "{}.values: must be non-empty",
                    path
                )));
            }
            for (i, v) in values.iter().enumerate() {
                if let AllowedValue::Float(f) = v {
                    if !f.is_finite() {
                        return Err(ContractError::Invalid(format!(
                            "{}.values[{}]: must be finite",
                            path, i
                        )));
                    }
                }
            }
        }
        Rule::Unique { columns, .. } => {
            if columns.is_empty() {
                return Err(ContractError::Invalid(format!(
                    "{}.columns: must be non-empty",
                    path
                )));
            }
        }
        Rule::DistinctCount { min, max, .. } => check_int_range(*min, *max, path)?,
        Rule::ForeignKeyIntegrity { references, .. } => {
            if references.table.is_empty() {
                return Err(ContractError::Invalid(format!(
                    "{}.references.table: required",
                    path
                )));
            }
            if references.column.is_empty() {
                return Err(ContractError::Invalid(format!(
                    "{}.references.column: required",
                    path
                )));
            }
        }
        Rule::RowCount { min, max, .. } => check_int_range(*min, *max, path)?,
        Rule::CustomSql { sql, .. } => {
            if sql.trim().is_empty() {
                return Err(ContractError::Invalid(format!("{}.sql: required", path)));
            }
        }
        Rule::NotEmpty { .. } => {}
    }
    Ok(())
}

fn check_int_range(min: Option<i64>, max: Option<i64>, path: &str) -> Result<(), ContractError> {
    if min.is_none() && max.is_none() {
        return Err(ContractError::Invalid(format!(
            "{}: provide at least one of min, max",
            path
        )));
    }
    if let (Some(a), Some(b)) = (min, max) {
        if a > b {
            return Err(ContractError::Invalid(format!(
                "{}: min ({}) must be <= max ({})",
                path, a, b
            )));
        }
    }
    Ok(())
}

fn check_float_range(
    min: Option<f64>,
    max: Option<f64>,
    path: &str,
) -> Result<(), ContractError> {
    if min.is_none() && max.is_none() {
        return Err(ContractError::Invalid(format!(
            "{}: provide at least one of min, max",
            path
        )));
    }
    if let Some(v) = min {
        if !v.is_finite() {
            return Err(ContractError::Invalid(format!("{}.min: must be finite", path)));
        }
    }
    if let Some(v) = max {
        if !v.is_finite() {
            return Err(ContractError::Invalid(format!("{}.max: must be finite", path)));
        }
    }
    if let (Some(a), Some(b)) = (min, max) {
        if a > b {
            return Err(ContractError::Invalid(format!(
                "{}: min ({}) must be <= max ({})",
                path, a, b
            )));
        }
    }
    Ok(())
}

fn ident_regex() -> Regex {
    Regex::new(r"^[A-Za-z_][A-Za-z0-9_]*$").expect("static regex")
}

fn duration_regex() -> Regex {
    Regex::new(r"^\d+(ms|s|m|h|d)$").expect("static regex")
}

/// Loose envelope used to extract the `name` from a contract file without
/// parsing the rest. Useful for indexing the contracts directory.
#[derive(Deserialize)]
struct NameOnly {
    name: String,
}

pub fn extract_name(source: &str) -> Option<String> {
    serde_yml::from_str::<NameOnly>(source)
        .ok()
        .map(|n| n.name)
        .filter(|n| ident_regex().is_match(n))
}

#[cfg(test)]
mod tests {
    use super::*;

    const VALID_YAML: &str = r#"
name: orders_quality
version: 1
target:
  connection: prod-pg
  schema: public
  table: orders
rules:
  - id: id_unique
    type: unique
    columns: [id]
  - id: status_enum
    type: allowed_values
    column: status
    values: [pending, paid, shipped, refunded]
  - id: amount_positive
    type: numeric_range
    column: amount_cents
    min: 0
    inclusive_min: true
  - id: customer_fk
    type: foreign_key_integrity
    column: customer_id
    references: { table: customers, column: id }
"#;

    #[test]
    fn parses_canonical_yaml() {
        let c = parse_contract(VALID_YAML, Format::Auto).expect("valid");
        assert_eq!(c.name, "orders_quality");
        assert_eq!(c.version, 1);
        assert_eq!(c.target.connection, "prod-pg");
        assert_eq!(c.rules.len(), 4);
    }

    #[test]
    fn parses_json_form() {
        let json = serde_json::json!({
            "name": "x",
            "version": 1,
            "target": { "connection": "c", "table": "t" },
            "rules": [{ "id": "r1", "type": "not_empty", "column": "c1" }],
        })
        .to_string();
        let c = parse_contract(&json, Format::Auto).expect("valid");
        assert_eq!(c.name, "x");
    }

    #[test]
    fn rejects_empty_source() {
        let err = parse_contract("   ", Format::Auto).unwrap_err();
        assert!(matches!(err, ContractError::Invalid(_)));
    }

    #[test]
    fn rejects_bad_name() {
        let src = r#"
name: "1bad-name"
version: 1
target: { connection: c, table: t }
rules: [{ id: r1, type: not_empty, column: c1 }]
"#;
        let err = parse_contract(src, Format::Auto).unwrap_err();
        assert!(err.to_string().contains("$.name"));
    }

    #[test]
    fn rejects_unsupported_version() {
        let src = r#"
name: ok
version: 2
target: { connection: c, table: t }
rules: [{ id: r1, type: not_empty, column: c1 }]
"#;
        let err = parse_contract(src, Format::Auto).unwrap_err();
        assert!(err.to_string().contains("version"));
    }

    #[test]
    fn rejects_duplicate_rule_id() {
        let src = r#"
name: ok
version: 1
target: { connection: c, table: t }
rules:
  - { id: dup, type: not_empty, column: a }
  - { id: dup, type: not_empty, column: b }
"#;
        let err = parse_contract(src, Format::Auto).unwrap_err();
        assert!(err.to_string().contains("duplicate"));
    }

    #[test]
    fn rejects_threshold_out_of_range() {
        let src = r#"
name: ok
version: 1
target: { connection: c, table: t }
rules:
  - { id: r1, type: not_null_pct, column: c1, threshold_min_pct: 150 }
"#;
        let err = parse_contract(src, Format::Auto).unwrap_err();
        assert!(err.to_string().contains("threshold_min_pct"));
    }

    #[test]
    fn rejects_invalid_regex() {
        let src = r#"
name: ok
version: 1
target: { connection: c, table: t }
rules:
  - { id: r1, type: regex_match, column: c1, pattern: "[" }
"#;
        let err = parse_contract(src, Format::Auto).unwrap_err();
        assert!(err.to_string().contains("regex"));
    }

    #[test]
    fn rejects_inverted_range() {
        let src = r#"
name: ok
version: 1
target: { connection: c, table: t }
rules:
  - { id: r1, type: numeric_range, column: c1, min: 100, max: 0 }
"#;
        let err = parse_contract(src, Format::Auto).unwrap_err();
        assert!(err.to_string().contains("min"));
    }

    #[test]
    fn rejects_range_without_bounds() {
        let src = r#"
name: ok
version: 1
target: { connection: c, table: t }
rules:
  - { id: r1, type: row_count }
"#;
        let err = parse_contract(src, Format::Auto).unwrap_err();
        assert!(err.to_string().contains("at least one"));
    }

    #[test]
    fn rejects_bad_max_age() {
        let src = r#"
name: ok
version: 1
target: { connection: c, table: t }
rules:
  - { id: r1, type: date_range, column: c1, max_age: "7days" }
"#;
        let err = parse_contract(src, Format::Auto).unwrap_err();
        assert!(err.to_string().contains("max_age"));
    }

    #[test]
    fn rejects_empty_unique_columns() {
        let src = r#"
name: ok
version: 1
target: { connection: c, table: t }
rules:
  - { id: r1, type: unique, columns: [] }
"#;
        let err = parse_contract(src, Format::Auto).unwrap_err();
        assert!(err.to_string().contains("columns"));
    }

    #[test]
    fn rejects_empty_allowed_values() {
        let src = r#"
name: ok
version: 1
target: { connection: c, table: t }
rules:
  - { id: r1, type: allowed_values, column: c1, values: [] }
"#;
        let err = parse_contract(src, Format::Auto).unwrap_err();
        assert!(err.to_string().contains("values"));
    }

    #[test]
    fn round_trip_json() {
        let c = parse_contract(VALID_YAML, Format::Auto).expect("valid");
        let json = serde_json::to_string(&c).expect("ser");
        let back = parse_contract(&json, Format::Json).expect("re-parse");
        assert_eq!(c, back);
    }

    #[test]
    fn extract_name_works() {
        assert_eq!(
            extract_name(VALID_YAML).as_deref(),
            Some("orders_quality")
        );
        assert!(extract_name("garbage").is_none());
    }

    #[test]
    fn rule_id_and_enabled_accessors() {
        let c = parse_contract(VALID_YAML, Format::Auto).expect("valid");
        assert_eq!(c.rules[0].id(), "id_unique");
        assert!(c.rules[0].enabled());
    }
}

File: src/lib/contracts/parser.ts
// SPDX-License-Identifier: BUSL-1.1

/**
 * Contract parser — YAML/JSON source → validated `Contract`.
 *
 * Validates structure, required fields, and rule-type-specific invariants
 * (e.g. regex compiles, numeric_range has at least one bound).
 */

import { parse as parseYaml } from 'yaml';
import type {
  AllowedValue,
  Contract,
  ContractTarget,
  ForeignKeyReference,
  Rule,
  RuleType,
  Severity,
} from './types';

export class ContractParseError extends Error {
  readonly path: string;
  readonly cause?: Error;

  constructor(message: string, path: string = '$', cause?: Error) {
    super(path === '$' ? message : `${path}: ${message}`);
    this.name = 'ContractParseError';
    this.path = path;
    this.cause = cause;
  }
}

export type ContractFormat = 'yaml' | 'json' | 'auto';

const SEVERITIES: ReadonlySet<Severity> = new Set(['error', 'warning', 'info']);

const RULE_TYPES: ReadonlySet<RuleType> = new Set([
  'not_null_pct',
  'not_empty',
  'regex_match',
  'length_range',
  'numeric_range',
  'date_range',
  'allowed_values',
  'unique',
  'distinct_count',
  'foreign_key_integrity',
  'row_count',
  'custom_sql',
]);

const IDENT_RE = /^[A-Za-z_][A-Za-z0-9_]*$/;
const DURATION_RE = /^(\d+)(ms|s|m|h|d)$/;

export function parseContract(source: string, format: ContractFormat = 'auto'): Contract {
  if (typeof source !== 'string' || source.trim() === '') {
    throw new ContractParseError('Source is empty.');
  }

  const raw = decodeSource(source, format);
  return validateContract(raw);
}

function decodeSource(source: string, format: ContractFormat): unknown {
  if (format === 'json') return parseJsonStrict(source);
  if (format === 'yaml') return parseYamlStrict(source);

  const trimmed = source.trimStart();
  if (trimmed.startsWith('{') || trimmed.startsWith('[')) {
    try {
      return JSON.parse(source);
    } catch {
      // fall through to YAML — JSON is a YAML subset
    }
  }
  return parseYamlStrict(source);
}

function parseJsonStrict(source: string): unknown {
  try {
    return JSON.parse(source);
  } catch (err) {
    throw new ContractParseError(`Invalid JSON: ${(err as Error).message}`, '$', err as Error);
  }
}

function parseYamlStrict(source: string): unknown {
  try {
    return parseYaml(source, { prettyErrors: true });
  } catch (err) {
    throw new ContractParseError(`Invalid YAML: ${(err as Error).message}`, '$', err as Error);
  }
}

function validateContract(raw: unknown): Contract {
  const obj = expectObject(raw, '$');

  const name = expectString(obj.name, '$.name');
  if (!IDENT_RE.test(name)) {
    throw new ContractParseError(
      'Must match [A-Za-z_][A-Za-z0-9_]* (used as filename + identifier).',
      '$.name'
    );
  }

  const version = expectInt(obj.version, '$.version');
  if (version !== 1) {
    throw new ContractParseError(`Unsupported version ${version} (expected 1).`, '$.version');
  }

  const description = optionalString(obj.description, '$.description');
  const target = validateTarget(obj.target);

  const rulesRaw = obj.rules;
  if (!Array.isArray(rulesRaw) || rulesRaw.length === 0) {
    throw new ContractParseError('Must be a non-empty array.', '$.rules');
  }

  const seenIds = new Set<string>();
  const rules: Rule[] = rulesRaw.map((r, i) => {
    const rule = validateRule(r, `$.rules[${i}]`);
    if (seenIds.has(rule.id)) {
      throw new ContractParseError(`Duplicate rule id "${rule.id}".`, `$.rules[${i}].id`);
    }
    seenIds.add(rule.id);
    return rule;
  });

  return { name, version, description, target, rules };
}

function validateTarget(raw: unknown): ContractTarget {
  const obj = expectObject(raw, '$.target');
  const connection = expectString(obj.connection, '$.target.connection');
  const table = expectString(obj.table, '$.target.table');
  const schema = optionalString(obj.schema, '$.target.schema');
  return { connection, schema, table };
}

function validateRule(raw: unknown, path: string): Rule {
  const obj = expectObject(raw, path);
  const id = expectString(obj.id, `${path}.id`);
  if (!IDENT_RE.test(id)) {
    throw new ContractParseError('Must match [A-Za-z_][A-Za-z0-9_]*.', `${path}.id`);
  }
  const type = expectString(obj.type, `${path}.type`);
  if (!RULE_TYPES.has(type as RuleType)) {
    throw new ContractParseError(`Unknown rule type "${type}".`, `${path}.type`);
  }
  const description = optionalString(obj.description, `${path}.description`);
  const severity = optionalSeverity(obj.severity, `${path}.severity`);
  const enabled = optionalBool(obj.enabled, `${path}.enabled`);
  const base = { id, description, severity, enabled };

  switch (type as RuleType) {
    case 'not_null_pct': {
      const column = expectString(obj.column, `${path}.column`);
      const threshold_min_pct = expectNumber(obj.threshold_min_pct, `${path}.threshold_min_pct`);
      if (threshold_min_pct < 0 || threshold_min_pct > 100) {
        throw new ContractParseError('Must be in [0, 100].', `${path}.threshold_min_pct`);
      }
      return { ...base, type: 'not_null_pct', column, threshold_min_pct };
    }
    case 'not_empty': {
      const column = expectString(obj.column, `${path}.column`);
      return { ...base, type: 'not_empty', column };
    }
    case 'regex_match': {
      const column = expectString(obj.column, `${path}.column`);
      const pattern = expectString(obj.pattern, `${path}.pattern`);
      try {
        new RegExp(pattern);
      } catch (err) {
        throw new ContractParseError(`Invalid regex: ${(err as Error).message}`, `${path}.pattern`);
      }
      return { ...base, type: 'regex_match', column, pattern };
    }
    case 'length_range': {
      const column = expectString(obj.column, `${path}.column`);
      const min = optionalInt(obj.min, `${path}.min`);
      const max = optionalInt(obj.max, `${path}.max`);
      requireAtLeastOne(min, max, `${path}`, 'min', 'max');
      if (min !== undefined && max !== undefined && min > max) {
        throw new ContractParseError('min must be <= max.', path);
      }
      return { ...base, type: 'length_range', column, min, max };
    }
    case 'numeric_range': {
      const column = expectString(obj.column, `${path}.column`);
      const min = optionalNumber(obj.min, `${path}.min`);
      const max = optionalNumber(obj.max, `${path}.max`);
      requireAtLeastOne(min, max, `${path}`, 'min', 'max');
      if (min !== undefined && max !== undefined && min > max) {
        throw new ContractParseError('min must be <= max.', path);
      }
      const inclusive_min = optionalBool(obj.inclusive_min, `${path}.inclusive_min`);
      const inclusive_max = optionalBool(obj.inclusive_max, `${path}.inclusive_max`);
      return { ...base, type: 'numeric_range', column, min, max, inclusive_min, inclusive_max };
    }
    case 'date_range': {
      const column = expectString(obj.column, `${path}.column`);
      const min = optionalString(obj.min, `${path}.min`);
      const max = optionalString(obj.max, `${path}.max`);
      const max_age = optionalString(obj.max_age, `${path}.max_age`);
      if (min === undefined && max === undefined && max_age === undefined) {
        throw new ContractParseError('Provide at least one of min, max, max_age.', path);
      }
      if (max_age !== undefined && !DURATION_RE.test(max_age)) {
        throw new ContractParseError('Must be <number><ms|s|m|h|d>, e.g. "7d".', `${path}.max_age`);
      }
      return { ...base, type: 'date_range', column, min, max, max_age };
    }
    case 'allowed_values': {
      const column = expectString(obj.column, `${path}.column`);
      const values = expectArray(obj.values, `${path}.values`);
      if (values.length === 0) {
        throw new ContractParseError('Must be non-empty.', `${path}.values`);
      }
      const checked: AllowedValue[] = values.map((v, i) => {
        if (v === null) return null;
        const t = typeof v;
        if (t === 'string' || t === 'number' || t === 'boolean') return v as AllowedValue;
        throw new ContractParseError(
          'Only string/number/boolean/null allowed.',
          `${path}.values[${i}]`
        );
      });
      return { ...base, type: 'allowed_values', column, values: checked };
    }
    case 'unique': {
      const columns = expectArray(obj.columns, `${path}.columns`);
      if (columns.length === 0) {
        throw new ContractParseError('Must be non-empty.', `${path}.columns`);
      }
      const checked = columns.map((c, i) => expectString(c, `${path}.columns[${i}]`));
      return { ...base, type: 'unique', columns: checked };
    }
    case 'distinct_count': {
      const column = expectString(obj.column, `${path}.column`);
      const min = optionalInt(obj.min, `${path}.min`);
      const max = optionalInt(obj.max, `${path}.max`);
      requireAtLeastOne(min, max, `${path}`, 'min', 'max');
      if (min !== undefined && max !== undefined && min > max) {
        throw new ContractParseError('min must be <= max.', path);
      }
      return { ...base, type: 'distinct_count', column, min, max };
    }
    case 'foreign_key_integrity': {
      const column = expectString(obj.column, `${path}.column`);
      const references = validateFkRef(obj.references, `${path}.references`);
      return { ...base, type: 'foreign_key_integrity', column, references };
    }
    case 'row_count': {
      const min = optionalInt(obj.min, `${path}.min`);
      const max = optionalInt(obj.max, `${path}.max`);
      requireAtLeastOne(min, max, `${path}`, 'min', 'max');
      if (min !== undefined && max !== undefined && min > max) {
        throw new ContractParseError('min must be <= max.', path);
      }
      return { ...base, type: 'row_count', min, max };
    }
    case 'custom_sql': {
      const sql = expectString(obj.sql, `${path}.sql`);
      if (sql.trim() === '') {
        throw new ContractParseError('Must be non-empty.', `${path}.sql`);
      }
      return { ...base, type: 'custom_sql', sql };
    }
  }
}

function validateFkRef(raw: unknown, path: string): ForeignKeyReference {
  const obj = expectObject(raw, path);
  return {
    table: expectString(obj.table, `${path}.table`),
    column: expectString(obj.column, `${path}.column`),
    schema: optionalString(obj.schema, `${path}.schema`),
  };
}

function expectObject(v: unknown, path: string): Record<string, unknown> {
  if (v === null || typeof v !== 'object' || Array.isArray(v)) {
    throw new ContractParseError('Must be an object.', path);
  }
  return v as Record<string, unknown>;
}

function expectArray(v: unknown, path: string): unknown[] {
  if (!Array.isArray(v)) {
    throw new ContractParseError('Must be an array.', path);
  }
  return v;
}

function expectString(v: unknown, path: string): string {
  if (typeof v !== 'string' || v.length === 0) {
    throw new ContractParseError('Must be a non-empty string.', path);
  }
  return v;
}

function optionalString(v: unknown, path: string): string | undefined {
  if (v === undefined || v === null) return undefined;
  if (typeof v !== 'string') {
    throw new ContractParseError('Must be a string.', path);
  }
  return v;
}

function expectNumber(v: unknown, path: string): number {
  if (typeof v !== 'number' || !Number.isFinite(v)) {
    throw new ContractParseError('Must be a finite number.', path);
  }
  return v;
}

function optionalNumber(v: unknown, path: string): number | undefined {
  if (v === undefined || v === null) return undefined;
  return expectNumber(v, path);
}

function expectInt(v: unknown, path: string): number {
  const n = expectNumber(v, path);
  if (!Number.isInteger(n)) {
    throw new ContractParseError('Must be an integer.', path);
  }
  return n;
}

function optionalInt(v: unknown, path: string): number | undefined {
  if (v === undefined || v === null) return undefined;
  return expectInt(v, path);
}

function optionalBool(v: unknown, path: string): boolean | undefined {
  if (v === undefined || v === null) return undefined;
  if (typeof v !== 'boolean') {
    throw new ContractParseError('Must be a boolean.', path);
  }
  return v;
}

function optionalSeverity(v: unknown, path: string): Severity | undefined {
  if (v === undefined || v === null) return undefined;
  if (typeof v !== 'string' || !SEVERITIES.has(v as Severity)) {
    throw new ContractParseError('Must be "error", "warning", or "info".', path);
  }
  return v as Severity;
}

function requireAtLeastOne(
  a: unknown,
  b: unknown,
  path: string,
  nameA: string,
  nameB: string
): void {
  if (a === undefined && b === undefined) {
    throw new ContractParseError(`Provide at least one of ${nameA}, ${nameB}.`, path);
  }
}
